File: core/modules/log/Logger.cc
// -*- LSST-C++ -*-
/*
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
// See Logger.h

#include "log/Logger.h"

namespace {

/// Sink class responsible for synchronization.
class SyncSink : public boost::iostreams::sink {
public:
    SyncSink(std::ostream* os) : _os(os) {}
    std::streamsize write(const char *s, std::streamsize n) {
        boost::mutex::scoped_lock lock(_mutex);
        std::string message(s, n);
        *_os << message << std::flush;
        return n;
    }
private:
    std::ostream* _os;
    static boost::mutex _mutex;
};
boost::mutex SyncSink::_mutex;
static SyncSink syncSink(&std::cout);
static boost::iostreams::stream_buffer<SyncSink> syncBuffer(syncSink);
static std::ostream logStream(&syncBuffer);

} // Anonymous namespace

namespace lsst {
namespace qserv {
namespace log {

// Thread local storage to ensure one instance of Logger per thread.
boost::thread_specific_ptr<Logger> Logger::_instancePtr;

// Application-wide severity threshold.
Logger::Severity Logger::_severityThreshold = Info;
boost::mutex Logger::_mutex;


//Logger::SyncSink Logger::syncSink(&(std::cout));
//boost::iostreams::stream_buffer<Logger::SyncSink> Logger::syncBuffer(syncSink);
//std::ostream Logger::logStream(&syncBuffer);

////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////
// Filters
////////////////////////////////////////////////////////////////////////
/// Filter class responsible for enforcing severity level.
class Logger::SeverityFilter : public boost::iostreams::multichar_output_filter {
public:
    SeverityFilter(Logger* loggerPtr);
    template<typename Sink>
    std::streamsize write(Sink& dest, const char* s, std::streamsize n);
private:
    Logger* _loggerPtr;
};

/// Filter class responsible for formatting Logger output.
class Logger::LogFilter : public boost::iostreams::line_filter {
public:
    LogFilter(Logger* loggerPtr);
private:
    std::string do_filter(const std::string& line);
    std::string getTimeStamp();
    std::string getThreadId();
    std::string getSeverity();
    Logger* _loggerPtr;
};


////////////////////////////////////////////////////////////////////////
// public
////////////////////////////////////////////////////////////////////////
#if 0
Logger::SyncSink::SyncSink(std::ostream* os) : boost::iostreams::sink() {
    _os = os;
}
#endif

Logger& Logger::Instance() {
    if (_instancePtr.get() == NULL) _instancePtr.reset(new Logger);
    return *_instancePtr;
}

// Allows retrieval of singleton and setting of severity
// with single function call.
Logger& Logger::Instance(Logger::Severity severity) {
    if (_instancePtr.get() == NULL) _instancePtr.reset(new Logger);
    _instancePtr.get()->setSeverity(severity);
    return *_instancePtr;
}

void Logger::setSeverity(Logger::Severity severity) {
    if (severity != _severity) {
        flush();
        _severity = severity;
    }
}

Logger::Severity Logger::getSeverity() const {
    return _severity;
}

void Logger::setSeverityThreshold(Logger::Severity severity) {
    boost::mutex::scoped_lock lock(Logger::_mutex);
    if (severity != _severityThreshold) {
        flush();
        _severityThreshold = severity;
    }
}

Logger::Severity Logger::getSeverityThreshold() const {
    return _severityThreshold;
}

////////////////////////////////////////////////////////////////////////
// private
////////////////////////////////////////////////////////////////////////

Logger::Logger() : boost::iostreams::filtering_ostream() {
    _severity = Info;
    Logger::SeverityFilter severityFilter(this);
    Logger::LogFilter logFilter(this);
    push(severityFilter);
    push(logFilter);
    push(logStream);
}

Logger::SeverityFilter::SeverityFilter(Logger* loggerPtr)
    : boost::iostreams::multichar_output_filter() {
    _loggerPtr = loggerPtr;
}

template<typename Sink>
std::streamsize Logger::SeverityFilter::write(Sink& dest, const char* s, std::streamsize n) {
    if (_loggerPtr->getSeverity() >= _loggerPtr->getSeverityThreshold()) {
        std::streamsize z;
        for (z = 0; z < n; ++z) {
            if (!boost::iostreams::put(dest, s[z]))
                break;
        }
        return z;
    } else
        return n;
}

Logger::LogFilter::LogFilter(Logger* loggerPtr) : boost::iostreams::line_filter() {
    _loggerPtr = loggerPtr;
}

std::string Logger::LogFilter::do_filter(const std::string& line) {
    return getTimeStamp() + " " + getThreadId() + " " + getSeverity() + " " + line;
}

std::string Logger::LogFilter::getTimeStamp() {
    char fmt[64], buf[64];
    struct timeval tv;
    struct tm* tm;
    gettimeofday(&tv, NULL);
    struct tm newtime;
    memset(&newtime, 0, sizeof(struct tm));
    localtime_r(&tv.tv_sec, &newtime);
    strftime(fmt, sizeof fmt, "%Y%m%d %H:%M:%S.%%06u", &newtime);
    snprintf(buf, sizeof buf, fmt, tv.tv_usec);
    return std::string(buf);
}

std::string Logger::LogFilter::getThreadId() {
    return boost::lexical_cast<std::string>(boost::this_thread::get_id());
}

std::string Logger::LogFilter::getSeverity() {
    switch (_loggerPtr->getSeverity()) {
    case Logger::Debug:
        return "DBG";
    case Logger::Info:
        return "INF";
    case Logger::Warning:
        return "WRN";
    case Logger::Error:
        return "ERR";
    default:
        return NULL;
    }
}

}}} // namespace lsst::qserv::log

File: admin/python/lsst/qserv/admin/logger.py
#!/usr/bin/env python

# LSST Data Management System
# Copyright 2015 AURA/LSST.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.

"""
Logging module for Qserv administration tools

@author  Fabrice Jammes, IN2P3

"""

from __future__ import absolute_import, division, print_function

# --------------------------------
#  Imports of standard modules --
# -------------------------------
import logging
import logging.config
import os

# ----------------------------
# Imports for other modules --
# ----------------------------

# ---------------------------------
# Local non-exported definitions --
# ---------------------------------

# -----------------------
# Exported definitions --
# -----------------------


def get_default_log_conf():
    default_log_conf = "{0}/.lsst/logging.ini".format(os.path.expanduser('~'))
    return default_log_conf


def add_logfile_opt(parser):
    """
    Add option to command line interface in order to set path to standard
    configuration file for python logger
    Input object isn't duplicated here.
    @param parser: a parser which may contains some options (mutable object)
    @type: argparse.ArgumentParser
    @rtype : argparse.ArgumentParser
    """

    parser.add_argument("-V", "--log-cfg", dest="log_conf",
                        default=get_default_log_conf(),
                        help="Absolute path to file containing python" +
                        "logger standard configuration file")
    return parser


def setup_logging(path='logging.ini',
                  default_level=logging.INFO):
    """
    Setup logging configuration from yaml file
    if the yaml file doesn't exists:
    - return false
    - configure logging to default_level
    """
    if os.path.exists(path):
        with open(path, 'r') as f:
            logging.config.fileConfig(f)
        return True
    else:
        logging.basicConfig(level=default_level)
        return False


def init_default_logger(log_file_prefix, level=logging.DEBUG, log_path="."):
    if level == logging.DEBUG:
        fmt = '%(asctime)s {%(pathname)s:%(lineno)d} %(levelname)s %(message)s'
    else:
        fmt = '%(asctime)s %(levelname)s %(message)s'
    add_console_logger(level, fmt)
    logger = add_file_logger(log_file_prefix, level, log_path, fmt)
    return logger


def add_console_logger(level=logging.DEBUG, fmt='%(asctime)s %(levelname)s %(message)s'):
    logger = logging.getLogger()
    formatter = logging.Formatter(fmt)
    logger.setLevel(level)
    console_handler = logging.StreamHandler()
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)

    return logger


def add_file_logger(log_file_prefix, level=logging.DEBUG, log_path=".",
                    format='%(asctime)s %(levelname)s %(message)s'):

    logger = logging.getLogger()
    formatter = logging.Formatter(format)
    # this level can be reduce for each handler
    logger.setLevel(level)
    logfile = os.path.join(log_path, log_file_prefix + '.log')
    file_handler = logging.FileHandler(logfile)
    file_handler.setFormatter(formatter)
    logger.addHandler(file_handler)

    return logger
